src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserBase.cs

```diff
@@ -0,0 +1,180 @@
+        /// </summary>
+        internal char PathSeparator { get; set; }
+
+        /// <summary>
+        /// Returns the PathSeparator values of the Server.
+        /// </summary>
+        /// <returns>PathSeparator</returns>
+        internal static char GetPathSeparator(Enumerator enumerator, SqlConnection connection)
+        {
+            var req = new Request();
+            req.Urn = "Server";
+            req.Fields = new[] { "PathSeparator" };
+            string pathSeparator = string.Empty;
+
+            using (DataSet ds = enumerator.Process(connection, req))
+            {
+                if (ds.Tables[0].Rows.Count > 0)
+                {
+                    pathSeparator = Convert.ToString(ds.Tables[0].Rows[0][0], System.Globalization.CultureInfo.InvariantCulture);
+                }
+            }
+
+            if (string.IsNullOrEmpty(pathSeparator))
+            {
+                pathSeparator = @"\";
+            }
+
+            return pathSeparator[0];
+        }
+
+        /// <summary>
+        /// Enumerates the FileInfo objects associated with drives 
+        /// </summary>
+        /// <param name="enumerator"></param>
+        /// <param name="connection"></param>
+        /// <returns></returns>
+        internal static IEnumerable<FileInfo> EnumerateDrives(Enumerator enumerator, SqlConnection connection)
+        {
+            // if not supplied, server name will be obtained from urn
+            Request req = new Request();
+            bool clustered = false;
+
+            req.Urn = "Server/Information";
+            req.Fields = new string[] { "IsClustered", "PathSeparator", "HostPlatform" };
+
+            var pathSeparator = @"\";
+            var hostPlatform = HostPlatformNames.Windows;
+            try
+            {
+                using (DataSet ds = enumerator.Process(connection, req))
+                {
+                    if (ds.Tables[0].Rows.Count > 0)
+                    {
+                        clustered = Convert.ToBoolean(ds.Tables[0].Rows[0][0],
+                            CultureInfo.InvariantCulture);
+                        pathSeparator = Convert.ToString(ds.Tables[0].Rows[0][1], CultureInfo.InvariantCulture);
+                        hostPlatform = Convert.ToString(ds.Tables[0].Rows[0][2], CultureInfo.InvariantCulture);
+                    }
+                }
+            }
+            catch (UnknownPropertyEnumeratorException)
+            {
+                //there can be no clusters on 7.0 server
+            }
+
+            // we need to issue different queries to get all fixed drives on a normal server, and
+            // shared drives on a cluster
+            req.Urn = clustered ? "Server/AvailableMedia[@SharedDrive=true()]" : "Server/Drive";
+            req.Fields = new[] { "Name" };
+
+            using (DataSet ds = enumerator.Process(connection, req))
+            {
+                for (int i = 0; i < ds.Tables[0].Rows.Count; i++)
```

Comment:
>ds.Tables[0] [](start = 36, length = 12)

do we need to check that Tables has a Count > 0?  I won't comment all ever reference of DataSet instances in this review, but as a general comment please review them all to make sure we are doing appropriate null checking and index bounds checking, etc.

Reply:
In SSMS code, it's only checking if the ds.Tables[0].Rows has items. And from my own testing, as long as we provide the supported filed types, we always get ds.Tables[0]. If there's no returned value, then it won't have any rows. However, looking at the enumerator.Process code, it could return null, so I added safe null and index bound check.